interp/constrintern.ml

```diff
@@ -1141,9 +1141,16 @@ let check_number_of_pattern loc n l =
   if not (Int.equal n p) then raise (InternalizationError (loc,BadPatternsNumber (n,p)))
 
 let check_or_pat_variables loc ids idsl =
-  if List.exists (fun ids' -> not (List.eq_set (fun {loc;v=id} {v=id'} -> Id.equal id id') ids ids')) idsl then
-    user_err ?loc  (str
-    "The components of this disjunctive pattern must bind the same variables.")
+  let eq_id {v=id} {v=id'} = Id.equal id id' in
+  let idsl = List.filter (fun ids' -> not (List.eq_set eq_id ids ids')) idsl in
+  match idsl with
+  | ids'::_ ->
+    let ids'' = List.subtract eq_id ids ids' in
+    let ids'' = if ids'' = [] then List.subtract eq_id ids' ids else ids'' in
+    user_err ?loc
+      (strbrk "The components of this disjunctive pattern must bind the same variables (" ++
+       Id.print (List.hd ids'').v ++ strbrk " is not bound in all patterns).")
+  | [] -> ()
```

Comment:
It feels like the list manipulation functions here are high-level operations that should be properly abstracted and named, right?

Reply:
Indeed I mean to name both:
```
let idsl = List.filter (fun ids' -> not (List.eq_set eq_id ids ids')) idsl in
```
and
```
let ids'' = List.subtract eq_id ids ids' in
let ids'' = if ids'' = [] then List.subtract eq_id ids' ids else ids'' in
```
to my taste we have too many occurences of this style in the code; I'd be great to start commonalization/deduplication. Also, that would allow people to improve the implementation in the future.

I think it would be good to tweak it, but if you don't feel like doing let me know and I'll merge as-is.

Reply:
>  if you don't feel like doing let me know and I'll merge as-is.

Indeed, making such a cleaning properly would require time which is not prioritary in my short term.

Reply:
OK, done. Tell me if it is clear enough.

Reply:
Thanks a lot Hugo, merging as soon as CI passes.